daemon/discovery.go

```diff
@@ -87,32 +93,35 @@ func initDiscovery(backendAddress, advertiseAddress string, clusterOpts map[stri
 		backend: backend,
 		ticker:  time.NewTicker(heartbeat),
 		term:    make(chan bool),
+		readyCh: make(chan struct{}),
 	}
 	// We call Register() on the discovery backend in a loop for the whole lifetime of the daemon,
 	// but we never actually Watch() for nodes appearing and disappearing for the moment.
-	reloader.advertise(advertiseAddress)
+	go reloader.advertiseHeartbeat(advertiseAddress)
 	return reloader, nil
 }
 
-func (d *daemonDiscoveryReloader) advertise(address string) {
-	d.registerAddr(address)
-	go d.advertiseHeartbeat(address)
-}
-
-func (d *daemonDiscoveryReloader) registerAddr(addr string) {
-	if err := d.backend.Register(addr); err != nil {
-		log.Warnf("Registering as %q in discovery failed: %v", addr, err)
-	}
-}
-
 // advertiseHeartbeat registers the current node against the discovery backend using the specified
 // address. The function never returns, as registration against the backend comes with a TTL and
 // requires regular heartbeats.
 func (d *daemonDiscoveryReloader) advertiseHeartbeat(address string) {
+	var readyChanClosed bool
+
+	// Setup a short ticker until the first heartbeat has succeeded
+	t := time.NewTicker(500 * time.Millisecond)
 	for {
 		select {
-		case <-d.ticker.C:
-			d.registerAddr(address)
+		case <-t.C:
+			if err := d.backend.Register(address); err != nil {
+				log.Warnf("Registering as %q in discovery failed: %v", address, err)
+			} else {
+				if !readyChanClosed {
+					close(d.readyCh)
+					readyChanClosed = true
+					t.Stop()
+					t = d.ticker
```

Comment:
If the store is unreachable during the daemon restart (for any reason), will the ticker will be aggressively set for 500 msec for ever ? I think we should a timeout after which, it should go to normal default/user-configured heartbeat timer.